### DISCUSSION TITLE
Problem with python -m fitz extract fonts

### DISCUSSION THREAD

**Discussion by jwes44** [pymupdf/PyMuPDF#2849] [Looking for help]
When I extract fonts from a document and try to embed them in HTML, sometimes I can't make them work. I don't know if there is something I need to do or if there is something wrong in the extracted font. See ANCHJC+ChineseLiShu1-29.ttf in the attached file for an example. Fontforge shows the font name to be UntitledTTF.
Also, fitz adds a number to the filename to the file name, e.g. -29 above, which means I need to do extra steps to associate the file name to the name in the document. Can this be disabled?
[test.zip](https://github.com/pymupdf/PyMuPDF/files/13504256/test.zip)

**Comment by JorjMcKie** [comment]
Fonts embedded in a PDF are usually subset fonts (if they are derived from some font file). This is also the case here. You can recognize this by the 6 uppercase letters and a "+" sign prefixing the original name, `ANCHJC+`.

Building the subset must of course be done by some software - which of course may contain errors.
In any case, the resulting font is not complete: the internal font name may have been omitted (on purpose or by sloppiness), but the main reason is to leave out all glyphs that are not used by the text of the PDF. This can save a **_lot_** of file size - especially for Asian fonts.

So you cannot use such a font to write characters that have been deleted from the original font.
In addition:
The original glyph numbers (which the PDF creator used to map his Unicode character to that graphical representation) may not be the same as another program (like a browser) expects them to be.

So all-in-all there are quite a number of possible reasons why things may not work as you would like to.

The **_file name_** used by MuPDF to store the font on disk has nothing to do with any of the above considerations and is not the reason for problems.

**Reply by jwes44** [reply]
ANCHJC+ChineseLiShu1-29.ttf (and ANCHOE+NSimSun-36.ttf, though it is not used in the page I extracted and I wonder why It is included) are not valid fonts according to the Windows Font Viewer (and other programs), though the others are. All PDF viewers I have tried are able to display the page correctly. The question is: Why are some extracted fonts not valid and can they be fixed?

I am trying to convert a PDF file to html so the embedded fonts have the glyphs I need.

The name is only an inconvenience, as I am not able to derive the name directly from span[font], and have to do some processing to find the file.

**Reply by JorjMcKie** [reply]
Maybe this approach helps you:
```python
fitz.TOOLS.set_subset_fontnames(True)  # force return of full font name
True
for b in page.get_text("dict",flags=fitz.TEXTFLAGS_TEXT)["blocks"]:
    for l in b["lines"]:
        for s in l["spans"]:
            print(s["font"])

            
ANCHDA+WadeGiles
ANCHHA+WadeGilesBold
ANCHHA+WadeGilesBold
ANCHHA+WadeGilesBold
ANCHHA+WadeGilesBold
ANCHHA+WadeGilesBold
ANCHHA+WadeGilesBold
ANCHDA+WadeGiles
ANCHIC+WadeGilesItalic
ANCHIC+WadeGilesItalic
```
If you then take into account that MuPDF appends the font's xref number to the name - according to the scheme `ABCDEF+fontname-xref.ext` you should have a fairly stable algorithm ...
```python
pprint(page.get_fonts())
[(20, 'ttf', 'TrueType', 'ANCHDA+WadeGiles', 'TT2', 'WinAnsiEncoding'),
 (23, 'ttf', 'TrueType', 'ANCHHA+WadeGilesBold', 'TT4', 'WinAnsiEncoding'),
 (26, 'ttf', 'TrueType', 'ANCHIC+WadeGilesItalic', 'TT6', 'WinAnsiEncoding'),
 (29, 'ttf', 'Type0', 'ANCHJC+ChineseLiShu1', 'TT7', 'Identity-H'),
 (33, 'ttf', 'TrueType', 'ANCHMD+TimesNewRoman', 'TT9', 'WinAnsiEncoding'),
 (36, 'ttf', 'Type0', 'ANCHOE+NSimSun', 'TT10', 'Identity-H'),
 (41, 'ttf', 'TrueType', 'ANCICE+WadeGilesNarrow', 'TT12', 'WinAnsiEncoding'),
 (44, 'ttf', 'TrueType', 'ANCIEE+ChineseLiShu1', 'TT13', 'WinAnsiEncoding')]
```

**Reply by jwes44** [reply]
I downloaded a font validator from [https://github.com/HinTak/Font-Validator/releases/tag/FontVal-2.1.6](url) and there seem to be some serious problems with the extracted fonts. Attached are two of the reports.
[ANCHOE+NSimSun-36.pdf](https://github.com/pymupdf/PyMuPDF/files/13530180/ANCHOE%2BNSimSun-36.pdf)
[ANCIEE+ChineseLiShu1-44.pdf](https://github.com/pymupdf/PyMuPDF/files/13530182/ANCIEE%2BChineseLiShu1-44.pdf)

**Reply by JorjMcKie** [reply]
Well, there is nothing that can be done about this, as I tried to explain above.

**Reply by jwes44** [reply]
Are the fonts just being copied from the pdf or are they being reformatted? What I am trying to understand is how mupdf can use these fonts to display text while some of the extracted fonts are unusable?

**Reply by JorjMcKie** [reply]
The font binaries are extracted from the PDF as is and not processed in any way (other than decompression).
The fonts are not "unusable". They obviously work ... as long as they are used within the context of this PDF - by MuPDF and any PDF viewer.
You simply must accept the fact that font sub-setting **_may_** result in a font object that works for the selected few characters in the PDF for which it has been created - but no longer as an external font.

Also consider this: All the PDF creation logic is no longer there - only its result in the form of PDF text objects - which are sub-strings of the page's `/Contents` wrapped by key words "BT" and "ET". What you find inside these sub-strings are glyph numbers which point to vector graphics code snippets inside the subset font binary (which ultimately produce the visual appearance).
There is no guarantee that a different program can safely deduct the correct glyph number from a given Unicode - even when that Unicode is among the ones supported by the font subset.
The latter is especially true for Type0 fonts like in your case. These are so-called composite fonts, which among other things require the presence of a CMAP (pointed to by the `/ToUnicode` key in the font object). This table is a mapping between glyph numbers and Unicodes. During font sub-setting, that CMAP is also changed.

**Reply by jwes44** [reply]
The Type0 fonts don't appear to be Truetype fonts at all, yet they are extracted as .ttf. Why is that?

**Comment by JorjMcKie** [comment]
> The Type0 fonts don't appear to be Truetype fonts at all, yet they are extracted as .ttf. Why is that?

Shouldn't happen. Have an example?

**Reply by jwes44** [reply]
Many. Here are some in test.pdf.
Page 0 (5, 'ttf', 'Type0', 'ANCHOE+NSimSun', 'TT10', 'Identity-H', 0)
Page 1 (19, 'ttf', 'Type0', 'ANCJGO+TSCFGuyinM5TT', 'TT14', 'Identity-H', 0)
Page 2 (30, 'ttf', 'Type0', 'ANCJIO+TSCUKaiMTT', 'TT15', 'Identity-H', 0)
Page 3 (66, 'ttf', 'Type0', 'BEBNHO+TSCUKaiMTT', 'TT7', 'Identity-H', 0)
Page 6 (139, 'ttf', 'Type0', 'BIHDNC+TSCUKaiMTT', 'TT5', 'Identity-H', 0)
Page 7 (156, 'ttf', 'Type0', 'BIHDNC+TSCUKaiMTT', 'TT5', 'Identity-H', 0)
Page 8 (179, 'ttf', 'Type0', 'BIIBLP+SILDoulosIPA', 'TT8', 'Identity-H', 0)

[test.pdf](https://github.com/pymupdf/PyMuPDF/files/13665759/test.pdf)

**Reply by JorjMcKie** [reply]
They are all in TTF format. Either directly or based on descendant fonts where the glyphs are written using TTF technology.
MuPDF also extracts them all as ttf: `mutool extract test.pdf`.
Contact the MuPDF experts in [this](https://discord.gg/CyjPq8br) Discord channel to discuss any questions.
